### ISSUE TITLE
consistent-function-scoping false positive in loops

### ISSUE THREAD

**Issue by klimashkin** [sindresorhus/eslint-plugin-unicorn#792]
Hi!

`consistent-function-scoping` is complaining about a function that is being created in a loop, and that depends on a dynamic variable inside that loop.

```js
export const myMap = new Map();

for (const [value] of Object.values(values)) {
  const regexp = new RegExp(`^${value}\\/?$`);
  const match = href => regexp.test(href);

  myMap.set(value, {value, regexp, match});
}
```

```
ESLint: Move arrow function `match` to the outer scope.(unicorn/consistent-function-scoping)
```

**Comment by fisker** [comment]
I tried to fix this issue, but seems not easy as I thought. 

Other broken cases

```js
			for (const value in values) {
				const a = 1;
				const match = () => a;
			}
```

```js
			for (let i = 0; i < 9; i ++) {
				const a = 1;
				const match = () => a;
			}
```

**Comment by GuestrixErik** [comment]
@fisker I was just about to post a bug report on this, when I found your report.

I can confirm this is a problem. My (sadly less minimal) repro:

```ts
export const foo = (foos: readonly string[]): void => {
  for (const blockScopedConstant of foos) {
    // ✅ Nothing is suggested here, and everything is fine. ✅
    const fineClosure = (): void => {
      // 👇️ `fineClosure()` captures `blockScopedConstant` (and so couldn't move up).
      console.log(blockScopedConstant);
    };

    // 👇️ This renaming *shouldn't* make any difference (but it does).
    const sameBlockScopedConstant = blockScopedConstant;

    // ❌ Impossible suggestion: `Move arrow function 'buggyClosure' to the outer scope.` ❌
    const buggyClosure = (): void => {
      // 👇️ `buggyClosure()` captures `sameBlockScopedConstant` (and so couldn't move up).
      console.log(sameBlockScopedConstant);
    };
  }
};
```

Perhaps the only thing it contributes is the simple observation that aliasing `blockScopedConstant` triggers the bug (in this example).
